daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -885,8 +885,10 @@ class DirectOrBufferedDataOutputStream private[io] (
               bufOS.getFile.delete()
           }
           nBitsPut / 8
-        } else
-          directDOS.putBytes(bufOS.getBuf, 0, nBytes, finfo)
+        } else {
+          Assert.usage(nBytes <= Int.MaxValue)
```

Comment:
I think this shoudl be an ``Assert.invariant``? My understanding is that ``Assert.usage`` has the implication that we're just not following our own assumptions about a function or something. For example, a function can only be called if isError is false, so we might have ``Assert.usage(!isError)``. It's still possible to call a function when things are an error, but we shouldn't do that.

But ``Assert.invariant`` implies that it shouldn't even be possible, not matter what we do. In this case, the size of a non-file backed buffer should never be larger than IntMax because a non-file backed buffer is backed by an ArrayBufffer, which can never be larger than IntMax. So nBytes being larger than IntMax should be impossible.

Also, below this in the else block which I think starts with the exact same code with just different variable names (and missing the new Assert.usage you've added). While we are here, I wonder it it makes sense to move this chunk of code that writes whole bytes out of the if-else blocks and have it always run so we remove this duplicate logic. Then there's just the if-statement forhandling fragment bytes if they exist?

Reply:
So funnily enough, I originally had it as an Assert.invariant, but then I saw elsewhere in the code where we were doing the same Int.MaxValue check, it was using Assert.usage and I changed my code to match.  I'll update my code to be invariant, but would you want me to update the other assert too?

I'll also remove the duplicate code too, missed that when I was working on this yesterday.

Reply:
I assume you're talking about:
```scala
abstract class LocalBuffer[T <: java.nio.Buffer] {

  def getBuf(length: Long) = {
    Assert.usage(length <= Int.MaxValue)
    ...
  }
}
```
In this case, I think Assert.usage is correct because it's enforcing an assumption about the parameter value, and ensuring uses of this function don't break this assumption. So Assert.usage is sortof saying "XYZ shouldn't be used this way", whereas Assert.invariant says "XYZ shouldn't be possible". It's fairly pretty subtle difference, and I think these two checks are slightly different.

Reply:
Assert.usage is for incorrect usage of the API of a method/function, i.e., the caller is not "using" it correctly. Generally this is for checking args, but also checking that state is proper for the use of the method. A good example is next/hasNext. You are not supposed to call next() unless hasNext() is true. So a next() implementation can do Assert.usage(this.hasNext()).  Assert.invariant is just stating that at this point in the code this condition is an invariant, i.e., it is supposed to be true. It is more things that are deeper than arg or pre-condition checking of a call. One writes Assert.invariant in order to make clear the assumptions of the code, particularly those that are subtle, and so might get changed accidently as code is maintained.

Reply:
Makes sense.